src/simplifier.cpp

```diff
@@ -766,6 +768,37 @@ static void rankEdgeCollapses(Collapse* collapses, size_t collapse_count, const
 	}
 }
 
+static void rankEdgeCollapsesWithMach(Collapse* collapses, size_t collapse_count, const Vector3* vertex_positions, const float* vertex_mach, float mach_bound256, const Quadric* vertex_quadrics, const unsigned int* remap)
+{
+	for (size_t i = 0; i < collapse_count; ++i)
+	{
+		Collapse& c = collapses[i];
+
+		unsigned int i0 = c.v0;
+		unsigned int i1 = c.v1;
+
+		// most edges are bidirectional which means we need to evaluate errors for two collapses
+		// to keep this code branchless we just use the same edge for unidirectional edges
+		unsigned int j0 = c.bidi ? i1 : i0;
+		unsigned int j1 = c.bidi ? i0 : i1;
+
+		const Quadric& qi = vertex_quadrics[remap[i0]];
+		const Quadric& qj = vertex_quadrics[remap[j0]];
+
+		float ei = quadricError(qi, vertex_positions[i1]);
+		float ej = quadricError(qj, vertex_positions[j1]);
+
+		if(fabsf(vertex_mach[i1] - vertex_mach[j1]) >= mach_bound256) {
```

Comment:
This is the main difference - when mach values at opposite edge points pass a given threshold between points we set error to a high value which prevents collapse, looks good.

Reply:
Idea - if we wanted to collapse the mesh further we could try checking whether the parameter-wise second derivatives (rate of gradient change) are non-zero - for values close to zero we could collapse the edges anyway and handle color changing from the shader via simple linear interpolation.